fgpyo/io/io.py

```diff
@@ -0,0 +1,178 @@
+    assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+
+    if parent_must_exist:
+        parent: str = f"{path.parent.absolute()}"
+        assert Path(
+            parent
+        ).exists(), f"Cannot write file because parent diretory does not exist: {path}"
+        assert Path(
+            parent
+        ).is_dir(), f"Cannot write file because parent exists and is not a directory: {path}"
+        assert os.access(
+            parent, os.W_OK
+        ), f"Cannot write file because parent directory is not writeable: {path}"
+
+
+def to_reader(
+    path: Path, mode: str = "rt"
+) -> Union[gzip.GzipFile, io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading with a specifeied mode or default 'rt'.
+
+    Args:
+        path: Path to be read from
+        mode: Mode to open reader, default 'rt'
+
+    Example:
+    >>> reader = io.to_reader(path = Path("reader.txt"))
+    >>> reader.readlines()
+    >>> reader.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return gzip.open(path, mode=mode)
+    else:
+        return path.open(mode=mode)
+
+
+def to_writer(path: Path, mode: str = "wt") -> Union[gzip.GzipFile, IO[Any], io.TextIOWrapper]:
+    # TODO find compatible return type
+    """Opens a Path for reading and based on extension uses open() or gzip.open()
+
+    Args:
+        path: Path to write to
+        mode: Mode to open writer, default 'wt'
+
+    Example:
+    >>> writer = io.to_writer(path = Path("writer.txt"))
+    >>> writer.write(f"{something}\n")
+    >>> writer.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return gzip.open(path, mode=mode)
+    else:
+        return path.open(mode=mode)
+
+
+def read_lines(path: Path, strip: bool = True) -> Any:
+    """Takes a path and reads it into a list of strings, removing line terminators
+    along the way.
+    # TODO Make line strip optional
+
+    Args:
+        path: Path to read from
+        strip: Boolean to strip lines or not
+
+    Example:
+    >>> read_back: List[str] = io.read_lines(path)
+    """
+    reader = to_reader(path=path)
+    list_of_lines = reader.readlines()
+    reader.close()
+    if strip:
+        return [str(line).strip() for line in list_of_lines]
+    else:
+        return [str(line).replace("\n", "") for line in list_of_lines]
+
+
+def write_lines(path: Path, lines_to_write: Iterable[Any]) -> None:
+    """Writes a file with one line per item in provided iterable
+
+    Args:
+        path: Path to write to
+        lines_to_write: items to write to file
+
+    Example:
+    >>> lines: List[Any] = ["things to write", 100]
+    >>> path: Path = Path("file_to_write_to.txt")
+    >>> io.write_lines(path = path, lines_to_write = lines)
+    """
+    writer = to_writer(path=path)
+    # cast is needed to pass mypy
+    lines = cast(list, [f"{item}\n" for item in lines_to_write])
```

Comment:
Do we really need cast
```suggestion
    lines: List[str] = [f"{item}\n" for item in lines_to_write]
```

Also, do you really need to make a list of all the lines?  That could be memory intensive.  You have an iterator as input, so just write them one by one

Reply:
From what I have found online, `cast` is needed when mypy complains about type errors (that really are not type errors) for example I had to use `cast` in `to_writer() & to_reader()` to wrap gzip in `TextIOWrapper`.